cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/ProgramControllerServiceAdapter.java

```diff
@@ -39,44 +36,52 @@
 public class ProgramControllerServiceAdapter extends AbstractProgramController {
 
   private static final Logger LOG = LoggerFactory.getLogger(ProgramControllerServiceAdapter.class);
-  private static final Logger USERLOG = Loggers.mdcWrapper(LOG, Constants.Logging.EVENT_TYPE_TAG,
-                                                           Constants.Logging.USER_LOG_TAG_VALUE);
+  private static final Logger USER_LOG = Loggers.mdcWrapper(LOG, Constants.Logging.EVENT_TYPE_TAG,
+                                                            Constants.Logging.USER_LOG_TAG_VALUE);
 
   private final Service service;
-  private final CountDownLatch serviceStoppedLatch;
+  private volatile boolean stopRequested;
 
   public ProgramControllerServiceAdapter(Service service, ProgramRunId programRunId) {
     super(programRunId);
     this.service = service;
-    this.serviceStoppedLatch = new CountDownLatch(1);
     listenToRuntimeState(service);
   }
 
   @Override
-  protected void doSuspend() throws Exception {
-
+  protected void doSuspend() {
+    // no-op
   }
 
   @Override
-  protected void doResume() throws Exception {
-
+  protected void doResume() {
+    // no-op
   }
 
   @Override
   protected void doStop() throws Exception {
-    if (service.state() != Service.State.TERMINATED && service.state() != Service.State.FAILED) {
-      LOG.debug("stopping controller service for program {}.", getProgramRunId());
+    LOG.debug("Stopping controller service for program {}.", getProgramRunId());
+    stopRequested = true;
+    long gracefulTimeoutMillis = getGracefulTimeoutMillis();
+    if (gracefulTimeoutMillis < 0) {
       service.stopAndWait();
-      LOG.debug("stopped controller service for program {}, waiting for it to finish running listener hooks.",
-                getProgramRunId());
-      serviceStoppedLatch.await(30, TimeUnit.SECONDS);
-      LOG.debug("controller service for program {} finished running listener hooks.", getProgramRunId());
+    } else {
+      gracefulStop(gracefulTimeoutMillis);
     }
+    LOG.debug("Controller service for program {} finished running listener hooks.", getProgramRunId());
+  }
+
+  /**
+   * Requests a graceful shutdown of the service with a timeout. Subclass can override this if it supports
+   * graceful termination with timeout.
+   */
+  protected void gracefulStop(long gracefulTimeoutMillis) {
+    service.stopAndWait();
```

Comment:
These 2 paths end up executing the same  `service.stopAndWait()` method in this class. I assume this behavior is overridden in child classes

Reply:
yes. This is more a default implementation.